adcm_aio_client/core/config/_base.py

```diff
@@ -0,0 +1,346 @@
+
+        return self._initial_data
+
+    @property
+    def _current_config_level(self: Self) -> dict:
+        return self._current_data["config"]
+
+    def _set_parameter_value(self: Self, names: LevelNames, value: ParameterValueOrNone) -> None:
+        data = self._ensure_data_prepared_for_change()
+        set_nested_config_value(config=data["config"], level_names=names, value=value)
+
+    # protocol is position-based now, so required to silence check
+    def _set_activation_attribute(self: Self, names: LevelNames, value: bool) -> None:  # noqa: FBT001
+        data = self._ensure_data_prepared_for_change()
+
+        attribute_name = level_names_to_full_name(names)
+
+        try:
+            data["adcmMeta"][attribute_name]["isActive"] = value
+        except KeyError as e:
+            message = (
+                f"Failed to change activation attribute of {attribute_name}: not found in meta.\n"
+                "Either income data is incomplete or callback for this function is prepared incorrectly."
+            )
+            raise RuntimeError(message) from e
+
+    def _ensure_data_prepared_for_change(self: Self) -> dict:
+        if self._changed_data is None:
+            self._changed_data = deepcopy(self._initial_data)
+
+        return self._changed_data
+
+    def _convert_payload_formated_json_fields_inplace(
+        self: Self, parameters_spec: dict, data: dict, prefix: LevelNames
+    ) -> None:
+        for key, value in data.items():
+            parameter_spec = self._get_parameter_spec(key, parameters_spec=parameters_spec)
+            level_names = (*prefix, key)
+            if parameter_spec.get("format") == "json":
+                set_nested_config_value(data, (key,), self._json_value_from_payload_format(value))
+                self._json_fields.add(level_names)
+            elif isinstance(value, dict) and self._parameter_is_group(parameter_spec):
+                self._convert_payload_formated_json_fields_inplace(
+                    parameters_spec=parameters_spec[key]["properties"], data=value, prefix=level_names
+                )
+
+    def _convert_json_fields_to_payload_format_inplace(self: Self, data: dict) -> None:
+        for json_field_name in self._json_fields:
+            change_nested_config_value(data, json_field_name, self._json_value_to_payload_format)
+
+    def _json_value_to_payload_format(self: Self, value: ParameterValueOrNone) -> str | None:
+        if value is None:
+            return None
+
+        return json.dumps(value)
+
+    def _json_value_from_payload_format(self: Self, value: str | None) -> ParameterValueOrNone:
+        if isinstance(value, str):
+            return json.loads(value)
+
+        return None
+
+
+# FOREIGN SECTION
+#
+# these functions are heavily inspired by configuration rework in ADCM (ADCM-6034)
+
+ROOT_PREFIX = "/"
+
+
+def set_nested_config_value[T](config: dict[str, Any], level_names: LevelNames, value: T) -> T:
```

Comment:
Do you think we should somehow highlight functions that change objects on the fly?